hashing/src/chunk_with_proof.rs

```diff
@@ -0,0 +1,322 @@
+use crate::indexed_merkle_proof::IndexedMerkleProof;
+
+use casper_types::{
+    allocate_buffer,
+    bytesrepr::{Bytes, FromBytes, ToBytes},
+};
+
+use crate::{error, Digest};
+
+/// Represents a chunk of data with attached proof.
+#[derive(PartialEq, Debug, schemars::JsonSchema, serde::Serialize, serde::Deserialize)]
+#[schemars(with = "String", description = "Hex-encoded hash digest.")]
+#[serde(deny_unknown_fields)]
+pub struct ChunkWithProof {
+    proof: IndexedMerkleProof,
+    chunk: Vec<u8>,
+}
+
+impl ToBytes for ChunkWithProof {
+    fn to_bytes(&self) -> Result<Vec<u8>, casper_types::bytesrepr::Error> {
+        let mut result = allocate_buffer(self)?;
+        result.append(&mut self.proof.to_bytes()?);
+        result.append(&mut Bytes::from(self.chunk.as_slice()).to_bytes()?);
+        Ok(result)
+    }
+
+    fn serialized_length(&self) -> usize {
+        self.proof.serialized_length() + self.chunk.serialized_length()
+    }
+}
+
+impl FromBytes for ChunkWithProof {
+    fn from_bytes(bytes: &[u8]) -> Result<(Self, &[u8]), casper_types::bytesrepr::Error> {
+        let ((proof, chunk), remainder) = <(IndexedMerkleProof, Bytes)>::from_bytes(bytes)?;
+        Ok((
+            ChunkWithProof {
+                proof,
+                chunk: chunk.into(),
+            },
+            remainder,
+        ))
+    }
+}
+
+impl ChunkWithProof {
+    #[cfg(test)]
+    /// 10 bytes for testing purposes.
+    pub const CHUNK_SIZE_BYTES: usize = 10;
+
+    #[cfg(not(test))]
+    /// 1 MiB
+    pub const CHUNK_SIZE_BYTES: usize = 1 << 20;
+
+    /// Constructs the [ChunkWithProof] that contains the chunk of data with the appropriate index
+    /// and the cryptographic proof.
+    /// Empty data is always represented as single, empty chunk and not as zero chunks.
+    pub fn new(data: &[u8], index: u64) -> Result<Self, error::MerkleConstructionError> {
+        let (proof, chunk) = if data.is_empty() {
+            (IndexedMerkleProof::new([Digest::hash(&[])], index)?, vec![])
+        } else {
+            (
+                IndexedMerkleProof::new(
+                    data.chunks(Self::CHUNK_SIZE_BYTES).map(Digest::hash),
+                    index,
+                )?,
+                data[Self::CHUNK_SIZE_BYTES * (index as usize)
+                    ..data
+                        .len()
+                        .min(Self::CHUNK_SIZE_BYTES * ((index as usize) + 1))]
+                    .to_vec(),
```

Comment:
```suggestion
                data.chunks(Self::CHUNK_SIZE_BYTES)
                    .nth(index as usize)
                    .unwrap()
                    .to_vec(),
```
Both this code and the code it replaces could panic if `index` was out of bounds.  However, the call to `IndexedMerkleProof::new()?` ensures `index` is valid here.

We should either document that (in case e.g. someone re-ordered this to `let (chunk, proof)`, making the code panicking), or preferably just also check the `index` while constructing `chunk` and exit with an error in that case.

Reply:
Explicit index validation added.